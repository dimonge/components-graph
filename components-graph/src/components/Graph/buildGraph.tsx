import * as d3 from "d3";
import { ILink, INode, LinksType, NodesType } from "../../types";

export const createSvg = (
  container: HTMLDivElement,
  width: number,
  height: number
) => {
  const svg = d3
    .select(container)
    .append("svg")
    .attr("width", width)
    .attr("height", height)
    .append("g");

  return svg;
};

export const createLink = (svg: any, links: LinksType) => {
  return svg
    .selectAll("line")
    .data(links)
    .enter()
    .append("line")
    .attr("stroke", "grey");
};

export const createNode = (svg: any, nodes: NodesType) => {
  return svg
    .selectAll("circle")
    .data(nodes)
    .enter()
    .append("circle")
    .attr("r", 20)
    .attr("fill", (d: any) => d.color);
};

export const createLabel = (svg: any, nodes: NodesType, searchText: string) => {
  return svg
    .selectAll("text")
    .data(nodes)
    .enter()
    .append("text")
    .text((d: any) => d.name)
    .attr("text-anchor", "middle")
    .attr("dominant-baseline", "central")
    .attr("fill", (d: any) => {
      if (
        searchText &&
        d.name.toLowerCase().indexOf(searchText.toLowerCase()) !== -1
      ) {
        return "yellow";
      } else {
        return null;
      }
    });
};

export const createScale = (
  nodes: NodesType,
  width: number,
  height: number
) => {
  const xScale = d3.scaleLinear().domain([1, 5]).range([50, width]);
  const yScale = d3
    .scaleLinear()
    .domain([0, d3.max(nodes, (d: any) => d.y)])
    .range([200, height]);

  return { xScale, yScale };
};

type generateGraphType = {
  svg: any;
  nodes: NodesType;
  links: LinksType;
  searchText: string;
  width: number;
  height: number;
};
const generateGraph = ({
  svg,
  nodes,
  links,
  searchText,
  width,
  height,
}: generateGraphType) => {
  const link = createLink(svg, links);
  const node = createNode(svg, nodes);
  const label = createLabel(svg, nodes, searchText);

  const { xScale, yScale } = createScale(nodes, width, height);
  const ticked = () => {
    node
      .attr("cx", function (d: any) {
        return xScale(d.evolution);
      })
      .attr("cy", function (d: any) {
        return d.y;
      });

    link
      .attr("x1", function (d: any) {
        return xScale(d.source.evolution);
      })
      .attr("y1", function (d: any) {
        return d.source.y;
      })
      .attr("x2", function (d: any) {
        return xScale(d.target.evolution);
      })
      .attr("y2", function (d: any) {
        return d.target.y;
      });

    label.attr("x", (d: any) => xScale(d.evolution)).attr("y", (d: any) => d.y);
  };

  d3.forceSimulation(nodes)
    .force(
      "link",
      d3
        .forceLink()
        .id((d: any) => d.id)
        .links(links)
    )
    .force("charge", d3.forceManyBody().strength(-1000))
    .force("collide", d3.forceCollide().strength(-400))
    .force(
      "x",
      d3.forceX().x((d: any) => {
        return xScale(d.evolution);
      })
    )
    .force(
      "y",
      d3.forceY().y((d: any) => {
        return yScale(d.y);
      })
    )
    .on("end", ticked);
};
export default function buildGraph(
  container: HTMLDivElement,
  linksData: LinksType,
  nodesData: NodesType,
  searchText: string
) {
  const links = linksData.map((link: ILink) => Object.assign({}, link));
  const nodes = nodesData.map((node: INode) => Object.assign({}, node));

  const containerRect = container.getBoundingClientRect();

  const width: number = containerRect.width;
  const height: number = containerRect.height;

  if (d3.select(container).select("svg")) {
    d3.select(container).select("svg").remove();
  }

  const svg = createSvg(container, width, height);
  generateGraph({ svg, nodes, links, searchText, width, height });
  return {
    destroy: () => {
      svg.remove();
    },
    nodes: () => {
      return svg.node();
    },
  };
}
