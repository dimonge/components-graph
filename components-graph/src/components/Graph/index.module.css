.container {
  border: 1px solid grey;
  margin: 1rem;
  height: 80vh;
}

@media only screen and (max-width: 320px) {
  .container {
    border: 0;
    margin: 0;
  }
}
