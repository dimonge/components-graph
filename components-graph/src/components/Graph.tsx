import * as React from "react";
import { IComponents } from "../types";
import buildGraph from "./buildGraph";
import styles from "./graph.module.css";

interface IGraph {
  data: IComponents;
  searchText: string;
}

const Graph = ({ data, searchText }: IGraph) => {
  const containerRef = React.useRef<HTMLDivElement>(null);

  React.useEffect(() => {
    let fn;

    if (containerRef.current) {
      const { destroy } = buildGraph(
        containerRef.current,
        data.links,
        data.nodes,
        searchText
      );
      fn = destroy;
    }
    return fn;
  }, [searchText]);

  return <div ref={containerRef} className={styles.container} />;
};

export default Graph;
