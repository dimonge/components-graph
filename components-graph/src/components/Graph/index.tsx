import * as React from "react";
import { IComponents } from "../../types";
import buildGraph from "./buildGraph";
import styles from "./index.module.css";

interface IGraph {
  data: IComponents;
  searchText: string;
}

const Graph = ({ data, searchText }: IGraph) => {
  const containerRef = React.useRef<HTMLDivElement>(null);

  React.useEffect(() => {
    let fn;

    if (containerRef.current) {
      const { destroy } = buildGraph(
        containerRef.current,
        data.links,
        data.nodes,
        searchText
      );
      fn = destroy;
    }
    return fn;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchText]);

  return <div ref={containerRef} className={styles.container} />;
};

export default Graph;
