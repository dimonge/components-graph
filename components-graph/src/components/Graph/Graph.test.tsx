import React from "react";
import Graph from "./index";
import renderer from "react-test-renderer";

const data = {
  nodes: [
    { id: 0, evolution: 3.0, x: 277, y: 10, color: "#70147a", name: "User" },
    {
      id: 1,
      evolution: 3.0,
      x: 677,
      y: 300,
      color: "#70147a",
      name: "Website",
    },
    { id: 2, evolution: 3.5, x: 591, y: 434, color: "#008142", name: "CMS" },
    { id: 3, evolution: 4.0, x: 495, y: 586, color: "#008142", name: "CI/CD" },
    {
      id: 4,
      evolution: 4.5,
      x: 625,
      y: 637,
      color: "#008142",
      name: "Infrastructure",
    },
    {
      id: 5,
      evolution: 2.5,
      x: 475,
      y: 728,
      color: "#008142",
      name: "CI & Infrastructure Support",
    },
    {
      id: 6,
      evolution: 3.5,
      x: 845,
      y: 424,
      color: "#ccccaa",
      name: "Uptime Monitoring",
    },
    {
      id: 7,
      evolution: 3.2,
      x: 774,
      y: 561,
      color: "#aaaaaa",
      name: "3rd Party Modules",
    },
    {
      id: 8,
      evolution: 1.5,
      x: 413,
      y: 396,
      color: "#eecccc",
      name: "Branding",
    },
    {
      id: 9,
      evolution: 1.0,
      x: 9066,
      y: 504,
      color: "#cceecc",
      name: "Design",
    },
    {
      id: 10,
      evolution: 2.5,
      x: 379,
      y: 5055,
      color: "#aaaaaa",
      name: "Templates",
    },
  ],
  links: [
    { source: 1, target: 0, cost: 30, distance: 2 },
    { source: 3, target: 5, cost: 11, distance: 3 },
    { source: 4, target: 5, cost: 19, distance: 5 },
    { source: 2, target: 3, cost: 1, distance: 9 },
    { source: 2, target: 4, cost: 10, distance: 15 },
    { source: 1, target: 2, cost: 1, distance: 5 },
    { source: 1, target: 6, cost: 1, distance: 7 },
    { source: 2, target: 7, cost: 1, distance: 3 },
    { source: 1, target: 8, cost: 15, distance: 9 },
    { source: 8, target: 9, cost: 1, distance: 5 },
    { source: 8, target: 10, cost: 7, distance: 11 },
    { source: 2, target: 10, cost: 1, distance: 5 },
  ],
};

describe("Component Graph", () => {
  it("show return graph snapshot", () => {
    const tree = renderer
      .create(<Graph data={data} searchText={""} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
