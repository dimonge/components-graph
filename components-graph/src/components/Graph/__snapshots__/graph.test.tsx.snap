// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Graph show return graph snapshot 1`] = `
<div
  className="container"
/>
`;
